database/insertData.js

```diff
@@ -288,6 +288,18 @@ const queries = [
                   }
                 }
               }
+              {
+                actionCode: "changeOutcome"
+                condition: { value: true }
+                trigger: ON_APPLICATION_SUBMIT
+                parameterQueries: {
+                  application_id: {
+                    operator: "objectProperties"
+                    children: [{ value: { property: "record_id" } }]
```

Comment:
I saw this object in the other PR but wasn't sure how it is receiving the application_id. Is this something done by the evaluator? I couldn't find where this object is being passed.

Reply:
I know, it's confusing isn't it. The object that gets passed in is the Payload object from the Trigger queue, in Triggers&Actions:
```
for (const action of actions) {
    for (const key in action.parameter_queries) {
      action.parameter_queries[key] = await evaluateExpression(action.parameter_queries[key], {
        objects: [payload],
        pgConnection: PostgresDB,
      })
    }
```
This expression tells the evaluator to look at the `record_id` field of that object, which in this case is the `application_id`, since it comes from a trigger on the `application` table.

Reply:
I see. But how does it know that `the record_id field of that object, which in this case is the application_id`?

Reply:
The `TriggerPayload` (object sent to Trigger notification listener) has the following fields:
```
{
  id: number (id of this trigger in the trigger_queue table)
  trigger: TriggerStatus (e.g OnApplicationSubmit)
  table: string (which table was triggered -- in this case "application"
  record_id: number (the id of the record that was triggered from the above table)
}
```
So it's basically sending the information about the database record (`table`, `record_id`) that had its `trigger` field updated and what the trigger was (`trigger`).

This payload object is sent as the object passed into to any `evaluateExpression` calls needed to evaluate the parameters of the Action when it runs. So the evaluator, in this case, returns the `record_id` of the first (and only) object in the `objects` field of parameters, which is the `id` of the application in the `application` table that was triggered. It knows the `record_id` will refer to the `application` table, because it's acting on the `OnApplicationSubmit` trigger.

Basically, everything that the Actions need must be deduced from this payload and are evaluated using the `parameter_queries` of each template action before being sent to the action plugin function.

Sorry, I know that's confusing. 🤯 There's probably a better way to make this all clearer. Does the diagram [here](https://github.com/openmsupply/application-manager-server/wiki/Triggers-and-Actions) help at all? Happy to talk this through with you in person when you're in and hopefully it'll make it clearer for the both of us, and maybe we can see potential improvements.

Reply:
Yes it helped. Thanks. Is just so much to remember! But I think I understood the idea now. Good explanation